### PR TITLE
Fix `LambdaInvokeFunctionOperator` payload parameter type

### LINKED ISSUE #32227: LambdaInvokeFunctionOperator expects wrong type for payload arg
### Apache Airflow version

2.6.2

### What happened

I instantiate LambdaInvokeFunctionOperator in my DAG. 

I want to call the lambda function with some payload. After following code example from official documentation, I created a dict, and passed its json string version to the operator:

```
d = {'key': 'value'}
invoke_lambda_task = LambdaInvokeFunctionOperator(..., payload=json.dumps(d))
```

When I executed the DAG, this task failed. I got the following error message:

```
Invalid type for parameter Payload, value: {'key': 'value'}, type: <class 'dict'>, valid types: <class 'bytes'>, <class 'bytearray'>, file-like object
```

Then I went to official boto3 documentation, and found out that indeed, the payload parameter type is `bytes` or file. See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/lambda/client/invoke.html

### What you think should happen instead

To preserve backward compatibility, The API should encode payload argument if it is str, but also accept bytes or file, in which case it will be passed-through as is.

### How to reproduce

1. Create lambda function in AWS
2. Create a simple DAG with LambdaInvokeFunctionOperator
3. pass str value in the payload parameter; use json.dumps() with a simple dict, as lambda expects json payload
4. Run the DAG; the task is expected to fail

### Operating System

ubuntu

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==7.3.0

### Deployment

Virtualenv installation

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by eladi99**
Closes: https://github.com/apache/airflow/issues/32227

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by eladi99** [comment]
This PR is meant to solve this issue I've opened: https://github.com/apache/airflow/issues/32227

InvokeLambdaFunctionOperator now accepts bytes or str for its payload argument.
If the value is str, it is converted into bytes using encode() method.

This will fix incompatibility with corresponding boto3 functionality while not breaking the API.

File: tests/system/providers/amazon/aws/example_lambda.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ def delete_lambda(function_name: str):
     invoke_lambda_function = LambdaInvokeFunctionOperator(
         task_id="invoke_lambda_function",
         function_name=lambda_function_name,
-        payload=json.dumps({"SampleEvent": {"SampleData": {"Name": "XYZ", "DoB": "1993-01-01"}}}),
+        payload=json.dumps({"SampleEvent": {"SampleData": {"Name": "XYZ", "DoB": "1993-01-01"}}}).encode(),
```
Comment (eladg95) [inline tests/system/providers/amazon/aws/example_lambda.py]: I wonder if this change is necessary. Do we want to encourage airflow users to send bytes (which is consistent with boto3) or leave it in the current way of sending str?

File: airflow/providers/amazon/aws/hooks/lambda_function.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ def invoke_lambda(
         invocation_type: str | None = None,
         log_type: str | None = None,
         client_context: str | None = None,
-        payload: str | None = None,
+        payload: bytes | None = None,
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/hooks/lambda_function.py]: This is breaking change. You cannot do that. That will break any user using the hook with `payload` as string

File: airflow/providers/amazon/aws/operators/lambda_function.py [hunk 3]
```diff
@@ -150,19 +150,23 @@ def __init__(
...
         self.log_type = log_type
         self.qualifier = qualifier
         self.invocation_type = invocation_type
         self.client_context = client_context
         self.aws_conn_id = aws_conn_id
 
+        if type(payload) == str:
+            payload = payload.encode()
+
+        self.payload = payload
```
Comment (vincbeck) [inline airflow/providers/amazon/aws/operators/lambda_function.py]: I would do this in the hook to preserve backward compatibility

File: tests/system/providers/amazon/aws/example_lambda.py (hunk 1 above)
Comment (vincbeck) [reply tests/system/providers/amazon/aws/example_lambda.py]: I would leave it as str. The operator/hook takes care of the encoding

**Review by vincbeck — CHANGES_REQUESTED** [review]

**Review by vincbeck — APPROVED** [review]

**Comment by vincbeck** [comment]
You need to update the unit tests. But otherwise, LGTM

File: airflow/providers/amazon/aws/hooks/lambda_function.py [hunk 4]
```diff
@@ -65,6 +65,9 @@ def invoke_lambda(
         :param payload: The JSON that you want to provide to your Lambda function as input.
         :param qualifier: AWS Lambda Function Version or Alias Name
         """
+        if type(payload) == str:
```
Comment (uranusjr) [inline airflow/providers/amazon/aws/hooks/lambda_function.py]:
```suggestion
        if isinstance(payload, str):
```

**Review by eladkal — CHANGES_REQUESTED** [review]
LGTM
Please add test coverage for this change

**Comment by eladi99** [comment]
@eladkal @vincbeck can you refer me to the tests that should be updated?
Should I add some new once, and if so -- where?

**Comment by eladkal** [comment]
> @eladkal @vincbeck can you refer me to the tests that should be updated? Should I add some new once, and if so -- where?

Sure.

The hook is tested with string payload
https://github.com/apache/airflow/blob/9d9b15989a02042a9041ff86bc7e304bb06caa15/tests/providers/amazon/aws/hooks/test_lambda_function.py#L28

you should add a new test / extend existed test to also test byte payload.
I think in this case probably just making the test accept multiple inputs (strings and bytes) using  `@pytest.mark.parametrize `will do the trick (you can find many examples for this in the source code)


The same goes for testing the operator class
https://github.com/apache/airflow/blob/38b901ec3f07e6e65880b11cc432fb8ad6243629/tests/providers/amazon/aws/operators/test_lambda_function.py#L77

**Comment by eladi99** [comment]
@eladkal pls take a another look

File: tests/providers/amazon/aws/operators/test_lambda_function.py [hunk 5]
```diff
@@ -106,12 +114,14 @@ def test_invoke_lambda(self, hook_mock):
         value = operator.execute(None)
 
         assert value == "data was read"
+
+        invoke_payload = payload.encode() if isinstance(payload, str) else payload
```
Comment (eladkal) [inline tests/providers/amazon/aws/operators/test_lambda_function.py]: This doesn't feel right?
There shouldn't be added logic in the test.

**Comment by eladkal** [comment]
Static checks are failing

**Comment by elad-galili-ka** [comment]
> 

Fixed

**Review by eladkal — APPROVED** [review]

**Pull request merged**